feat(howto): add platform development howto (#189)
anbox-cloud/discourse-docs: howto/anbox/develop-platform.md

> Anbox Cloud provides a platform SDK which allows the development of custom platform plugins for the Anbox runtime.

Reviewer: ```suggestion
Anbox Cloud provides a platform SDK that allows the development of custom platform plugins for the Anbox runtime.
```

We should also link to https://anbox-cloud.io/docs/ref/platforms and give a short explanation and maybe some examples on why you would want to add your own platform.

Author: Added

This guide assumes all steps are being run on a machine running Ubuntu 22.04 and hosting the [Anbox Cloud Appliance](https://discourse.ubuntu.com/t/install-appliance/22681).

> A platform module must be build on the same version of Ubuntu as the Anbox runtime. This means if you're using one of the Ubuntun 22.04 based Anbox images (e.g. `jammy:android12:arm64`) you must build on Ubuntu 22.04.

Reviewer: ```suggestion
A platform module must be built on the same version of Ubuntu as the Anbox runtime. This means if you're using one of the Anbox images based on Ubuntu 22.04 (for example, `jammy:android12:arm64`), you must build on Ubuntu 22.04.
```
I guess that means the Anbox runtime runs inside the Ubuntu container? Adding that might make it easier to understand; I had to read this three times to understand that we're talking about the images first and then about the appliance in the next sentence.

In case that you're not running the [Anbox Cloud Appliance](https://discourse.ubuntu.com/t/install-appliance/22681) on an Ubuntu 22.04 system, you can build the platform on a separate system, e.g. in a LXD or docker container or another machine.

> To get started you first need to clone the [Anbox Platform SDK](https://github.com/anbox-cloud/anbox-platform-sdk) git repository:

Reviewer: Currently, the whole page is one chunk of text, which is very hard to follow - so we'll need to break this up.
I think it should work to use steps here (So "1. Install, 2. Build the example, 3. Launch" and so on, each with a few paragraphs and commands under them). If each step gets too long, we could use sections instead, but I think a list of steps should be fine for now.

Author: Added

    git clone https://github.com/anbox-cloud/anbox-platform-sdk.git

Before you can build one of the examples, you have to install all necessary build dependencies:

>     sudo apt install -y --no-install-recommends cmake build-essential ninja-build dh-exec cmake-extras libgmock-dev libelf-dev

Reviewer: Just link to https://anbox-cloud.io/docs/ref/sdks#anbox-platform-sdk for installation.
(Might need to update the dependencies there?)

Author: Done

Author: Added

> The [Anbox Platform SDK](https://github.com/anbox-cloud/anbox-platform-sdk) comes with various examples which demonstrate different features. The following steps are using the `minimal` example. Alternatively you can choose the `nvidia` example if you're working with NVIDIA GPUs and want to see graphical output.

Reviewer: ```suggestion
The [Anbox Platform SDK](https://github.com/anbox-cloud/anbox-platform-sdk) comes with various examples that demonstrate different features. The following steps use the `minimal` example. Alternatively, choose the `nvidia` example if you're working with NVIDIA GPUs and want to see graphical output.
```

> To build the `minimal` platform example, run the following commands:

Reviewer: Should we point out to users that they should do some changes to the code first, or that this is the place where they would do their code changes?

Author: That's not needed to run the example as it's intended to run as-is.

Author: The example is meant to work as-is without any further changes.

    triplet=$(dpkg-architecture -qDEB_BUILD_MULTIARCH)
    cmake \
        -G Ninja \
        -B build \
        -DCMAKE_PREFIX_PATH=../../lib/anbox-platform-sdk/ \
        -DCMAKE_INSTALL_LIBDIR=lib/${triplet} \
        -DCMAKE_INSTALL_PREFIX=/usr examples/minimal/
    ninja -C build

As a result of the build process you will find a `platform_minimal.so` module in the `build` directory.

> [AMS](https://discourse.ubuntu.com/t/about-ams/24321) allows launching Anbox containers in a special development mode which is meant for development tasks, specifically for addon or platform development. The Anbox runtime is instructed, when in development mode, not to terminate the container when it detects failures or other problems.

Reviewer: Dev mode is documented here: https://anbox-cloud.io/docs/exp/containers#dev-mode

Author: Added

Author: Added

> To try out the minimal platform start a raw container with development mode turned on:
> 
>     amc launch --raw --devmode --instance-type=a4.3
> 
> If you choose the `nvidia` example you need to also selected an [instance type](https://discourse.ubuntu.com/t/instance-types/17764) which supports GPUs:
> 
>     amc launch --raw --devmode --instance-type=g4.3
> 
> As result, the command will print out the id of the container. Please take a note of the id as you will need it in the next step

Reviewer: ```suggestion
To try out the minimal platform, start a raw container with development mode turned on:

    amc launch --raw --devmode --instance-type=a4.3

If you choose the `nvidia` example, you must select an [instance type](https://discourse.ubuntu.com/t/instance-types/17764) that supports GPUs:

    amc launch --raw --devmode --instance-type=g4.3

The command will print out the ID of the container. Note down this ID; you will need it in the next step.
```

    id=<id of the container>
    triplet=$(dpkg-architecture -qDEB_BUILD_MULTIARCH)
    amc exec "${id}" mkdir "/usr/lib/${triplet}/anbox/platforms/minimal"
    lxc file push build/platform_minimal.so "ams-${id}/usr/lib/${triplet}/anbox/platforms/minimal/"

With the platform plugin present inside the container, you can now configure the Anbox runtime to make use of it. For that you have to change the `/var/lib/anbox/session.yaml` configuration file within the container:

    cat << EOF | amc exec "${id}" tee /var/lib/anbox/session.yaml
    log-level: debug
    platform: minimal
    EOF

This will rewrite the Anbox runtime configuration to use the new "minimal" platform instead of the default one.

As last step you have to restart Anbox inside the container

    amc exec "${id}" -- systemctl restart anbox

Once the Anbox runtime got restarted, you can verify from the container system log that it now uses the new "minimal" platform:

    amc logs "${id}"

Inside the log you will fine a line saying

    [ 2022-12-07 14:00:18] [anbox-session] [base_platform.cpp:34@create] Using platform 'minimal'

This shows that the platform was successfully loaded and is in use by Anbox.

> When doing development of a new platform you can also directly invoke the Anbox runtime rather than starting it via systemd. For that you first have to stop the `anbox.service` unit
> 
>     amc exec "${id}" -- systemctl stop anbox
> 
> Now you can start the Anbox runtime directl by running

Reviewer: ```suggestion
When developing a new platform, you can also directly invoke the Anbox runtime rather than start it via `systemd`. For that, stop the `anbox.service` unit first:

    amc exec "${id}" -- systemctl stop anbox

Then start the Anbox runtime directly by running:
```
